{DJAN-5} Margarita upgrade instructions

vkurup opened the pull request.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+
+   .. code-block:: bash
+
+      ~/dev$ mv fabfile.py fabfile.py.old
```
Should this be `mv myproject/fabfile.py myproject/fabfile.py.old`?

vkurup replied:
Yes, I waffled about which dir I should use as the working directory. I'm going to change everything to be from myproject's directory, which will make things consistent all the way through the document and also make it easier to copy/paste commands since they won't have "myproject" in it.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+
+   1. Make sure ``SALT_VERSION`` is ``2015.5.8``
+   2. Change ``env.master`` in both the ``staging()`` and ``production()`` functions to be
+      correct (get the value from fabfile.py)
```
`get the value from fabfile.py.old`

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+Upgrade Salt
+------------
+
+#. Uninstall salt. **Make sure you are using the new fabfile!**
```
We could use a note here that we're doing this on staging but it will need to be repeated with `s/staging/production/`, right?

dpoirier replied:
Before removing salt, we should grab a copy at least of `/etc/salt/minion`, so we know what roles were set previously.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+
+      fab staging setup_master
+
+#. Set up the salt minion.
```
Change the list of roles in this command to just those that your project is using, based on the `/etc/salt/minion` file that was on the server.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+     "devDependencies": {}
+   }
+
+Deploy!!!
```
Need to commit the changes to some branch, push it to github, then update staging.sls to deploy that branch temporarily.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+---------------
+
+This must be done because the new fabfile has removed the *secrets-syncing* logic, so unsuspecting
+developers **will likely** stomp on each others secrets. Encrypting cannot  be done ``setup_master``
```
missing word near the end of this sentence? maybe "until"?

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+
+   .. code-block:: bash
+
+      fab staging encrypt:DB_PASSWORD='superSecretPassword'
```
Show bash prompt at the beginning of this line?

dpoirier replied:
And in the examples below.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+      +    - staging
+         'environment:production':
+
+#. Commit and redeploy:
```
Commit, push, and redeploy.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+* ``newrelic_license_key`` must be capitalized. Some projects have a secret for
+  ``newrelic_license_key``, but the current margarita uses ``NEW_RELIC_LICENSE_KEY``
+* NewRelic settings may need adjusting, which you can do via environment variables (see the
+  margarita README for details)
```
I just opened an issue (https://github.com/caktus/developer-documentation/issues/21) to move that documentation from the Margarita README to the development docs where the other information about setting up to use Margarita states lives.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+
+      ufw allow salt
+
+  If salt is installed, that will turn open up the ports that salt needs. Our deploy does that too,
```
s/will turn/will in turn/?

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,467 @@
+  If salt is installed, that will turn open up the ports that salt needs. Our deploy does that too,
+  but if the firewall is already running, then our salt state can't run. I'm still confused how this
+  problem happened on a server which had already been running salt successfully, but ¯\\_(ツ)_/¯.
+* *VAGRANT NOTE*: Make sure to undo the setting which syncs the conf folder to ``/srv`` on the VM,
```
mention where this setting lives?

dpoirier commented:
Could this be re-organized to first update the project files to use the new Margarita, test by doing a clean deploy to Vagrant, and only then starting to update staging and then production?

vkurup commented:
Thanks for the great feedback, including the idea to do a clean deploy to Vagrant first. That way, all the code changes are hopefully done, so the staging and production updates are easier. Ready for re-review.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+   The new project template does not 'reconcile' secrets because it assumes that you have
+   committed the encrypted versions to the repo, so syncing back-and-forth is not necessary. We
+   haven't yet done that step for our project, so we need to be very careful about our secrets.
+   Make sure to sync them (by doing a deploy or running ``fab get_secrets``) before going any
```
Maybe we should be even more explicit and make this step 1?

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+
+#. Edit fabfile.py:
+
+   1. Make sure ``SALT_VERSION`` is ``2015.5.8``
```
Possible add a parenthetical ``(We know Margarita 1.4.0 works with Salt 2015.5.8.)``

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+
+   .. code-block:: yaml
+
+      margarita_version: 1.5.0
```
We said ``1.4.0`` above.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+      @@ -11,2 +11,5 @@ from celery import Celery
+
+      +from . import load_env
+      +load_env.load_env()
```
Does PEP-8 want blank lines between imports and code?

vkurup replied:
I guess not, this is how we have it in RescueID: https://github.com/theirc/rescue_id/blob/develop/rescue_id/celery.py#L5-L6

Should I add a line?

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+      @@ -4,2 +4,5 @@ import sys
+
+      +import dotenv
+      +dotenv.read_dotenv()
```
Why is this one different from the others?

vkurup replied:
This is the canonical usage recommended in the dotenv docs. The other ones are customized to work from subdirectories. See https://github.com/caktus/django-project-template/pull/208 for the PR where this was added.

dpoirier replied:
Our little helper ought to work from anywhere, so why not be consistent?  That would also avoid leaving around an example of code that someone might copy to use somewhere else without realizing it will only work right in the right directory.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+      +    (myproject)$ cp myproject/settings/local.example.py myproject/settings/local.py
+      +    (myproject)$ echo "DJANGO_SETTINGS_MODULE=myproject.settings.local" > .env
+
+
```
Possibly insert another step here to compare the current project template's README to the one in myproject and use our own judgment on any other updates we might want to make to myproject's README?

vkurup replied:
I mention that below, in step 4 under 'Miscellaneous Work'. Would you prefer it here? Or maybe in both places?

dpoirier replied:
True, I see now that you have a number of tasks like that at the end. I think that's fine, no change needed here.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+      less_version: 1.5.1   # or whatever you're currently using
+      postgres_version: 9.1 # or whatever you're currently using
+
+#. Move the old states out of the way (don't leave them in the conf directory), and copy in the
```
It might be a good idea before doing this to look for custom Salt states this project might have added. One way to approach that might be to compare the project's salt/top.sls to the template's one and see if the project has additional states, then see if Margarita has those. If not, we probably want to preserve those states as we update the project (and be sure to load them in the new salt/top.sls).

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+#. Get a copy of the ``Makefile`` from the project template, porting any functions in your existing
+   one to the new one, if needed.
+
+#. Review everything in ``salt.old`` to see which pieces are specific to your project and need to
```
If we preserved any custom states above, then we can simplify this slightly to just look to see what states remain in the project's salt directory and see if they can be moved to Margarita.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+   .. code-block:: bash
+
+      ~/dev/myproject$ fab vagrant setup_master
+      ~/dev/myproject$ fab -H 127.0.0.1:2222 vagrant setup_minion setup_minion:salt-master,web,worker,balancer,db-master,queue,cache
```
This invokes `setup_minion` twice.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+
+      ~/dev/myproject$ fab -H 1.2.3.4 staging -- sudo apt-get remove salt-master salt-minion salt-common -y --force-yes
+
+#. If you are on Ubuntu 12.04, run this command to enable backports, needed for python-gnupg:
```
I wonder why Margarita doesn't do this for us?

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+
+   .. code-block:: bash
+
+      ~/dev/myproject$ fab -H 1.2.3.4 staging -- sudo apt-get remove salt-master salt-minion salt-common -y --force-yes
```
BTW, I never knew we could invoke arbitrary commands on the server using fabric this way. Handy!

vkurup replied:
I know, I just learned about it while writing this documentation!

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+----
+
+#. Sync these states over to the server (do this separately from the actual deploy so that
+   failures can be caught before actually trying to deploy)
```
Missing period at the end of this sentence.

dpoirier commented on source/margarita/upgrading.rst:
```diff
@@ -0,0 +1,557 @@
+
+  and then restart the VM. If you are using the new ``Vagrantfile``, you shouldn't need to do that.
+
+* *VAGRANT NOTE:* Remove the ``source`` and ``public`` symlinks as we rsync now rather than symlink.
```
Consistency nit: here the ``:`` is included in the italicization, but in the previous bullet it isn't.

vkurup commented:
@dpoirier Updated. I think I addressed all the points but I wasn't quite sure if my notes about resolving custom states is sufficient. Your idea to search for differences in the top.sls file was a great one, so I've added it.

But I still think the upgrader will need to review each old state file completely (excluding the custom ones) to see if there are any changes to the states in margarita. As an example, epicallies added an additional port to be opened in the firewall. That state was edited in web/balancer.sls. So, just looking at state names, you'd never notice it.

Come to think of it, I should document what I did in that case. I found the commit after the first commit and used the Github 'compare changes' functionality to see all the code added since then. It was a huge comparison. Then in that diff, I looked only for changes in the conf directory. I can document that process, but wanted to see if you had any other feedback

dpoirier commented:
@vkurup looking good

vkurup commented:
I added the 'compare changes' docs. Want to take one last peek?

dpoirier commented:
:cocktail:
